ContactManager2/src/components/ContactForm.jsx

```diff
@@ -0,0 +1,43 @@
+import { InputTextBox } from "./InputTextBox";
+import { EMAIL, FIRST_NAME, LAST_NAME, contactFieldValidator } from "../config/contactFormConfig";
+import { useCallback, useState } from "react";
+import PropTypes from "prop-types"
+
+export const ContactForm = ({ createContact }) => {
+  const [formData, setFormData] = useState({});
+  const [errorMessages, setErrorMessages] = useState({});
+
+  const updateField = useCallback((id, value) => {
+    setFormData(data => ({
+      ...data, [id]: value
+    }))
+    contactFieldValidator[id](value, setErrorMessage)
+  }, [])
+
+  const setErrorMessage = useCallback((id, errorMessage) => {
+    setErrorMessages(messages => ({
+      ...messages, [id]: errorMessage
+    }))
+  }, [])
+
+  const submitAction = (event) => {
+    event.preventDefault();
+    const areAllFieldsValid = contactFieldValidator.validateAllFields(formData, setErrorMessage);
```

Comment:
no need to pass setErrorMessage as arguments... simply return error object from validators and then use setErrorMessage or do your computation by checking whether error object is empty or not.
error object will be like this - 
{firstname: "This field is required"}

Reply:
done